### PR TITLE
nixos/pppd: init

### PR DESCRIPTION
**Pull request by danderson**
###### Motivation for this change

pppd is used to control a lot of home internet connections. I want to use NixOS as my home router, and I need pppd to bring up my FTTH internet connection. It will also be useful to people using DSL (PPPoE or PPPoA) and classic dial-up internet.

pppd is already available in nix as the `ppp` derivation. It's used by other nixos modules like `pptpd`, but there's no module for running PPP sessions directly. This module fixes that.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

(Other tests)
- [x] Configured and ran a pppd service, and successfully got internet access using this module.

###### Notify maintainers

cc @falsifian (because you're the maintainer of the `ppp` package, which this module service-ifies)

### CODE REVIEW

**Comment by danderson** [comment]
One style question I didn't know how to answer: what should the commit message be? CONTRIBUTING.md says that it should be `init at <version>`, but this commit only adds a nixos module, which doesn't have a version.

**Comment by alexarice** [comment]
The commit title is fine

File: maintainers/maintainer-list.nix
Comment (alexarice) [inline maintainers/maintainer-list.nix]: The addition to `maintainer-list.nix` should be in a separate commit titled `maintainers: add danderson`

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: It is usual to put 
```
let
  cfg = config.services.pppd;
in
```
Here and then use that below

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: Is this necessary, I thought the document system could cope with this

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: `loaOf` should be being deprecated. See https://github.com/NixOS/nixpkgs/pull/63103 . I would use `attrsOf` here but if you might need multiple submodules per name (or if ordering matters) then use list

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: Think you need `default = name;` here or am I mistaken?

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: You could use `mkEnableOption` here

Comment (alexarice) [reply nixos/modules/services/networking/pppd.nix]: Ah you want it to be on by default, this is probably fine

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: Could consider using a structured config. See https://github.com/NixOS/rfcs/pull/42

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: I would prefer `default = name` as I have written above I think

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]:
```suggestion
  in mkIf cfg.enable {
```

I think you should include a global enable option

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: Should the package be added to the environment or is that not necessary (i.e if it is only run by systemd)

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: I don't know, I was copying other nixos modules that do this with their pkg reference. I'm happy to remove it and see what happens when the docs go live?

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: Yes, in this case I think the default should be true because it only matters if the submodule exists. So the most likely use case for this option is "I want to temporarily disable this 1 peer".

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: pppd configurations are a bit unusual compared to more modern configs, so I'm not sure if it fits cleanly inside a structured config. For example, here's my pppd config:

```
plugin rp-pppoe.so eno1
name "redacted-isp-username"
noipdefault
persist
defaultroute
hide-password
noauth
```

I'm not sure how I would convert this to a structured config, given the irregular shape. Do you have a suggestion?

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: I don't know what the custom is for NixOS. The only reason to invoke `pppd` directly is debugging a connection, but you can do that easily with `systemctl restart pppd-peer` as well. Based on that, I thought I would not pollute the system $PATH with a "useless" daemon, but it's not a strong opinion. WDYT?

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [reply nixos/modules/services/networking/pppd.nix]: It can't hurt so leave it

Comment (alexarice) [reply nixos/modules/services/networking/pppd.nix]: There is probably a way to locally build the docs but I'm not quite sure how if you wanted to check

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [reply nixos/modules/services/networking/pppd.nix]: I'm not sure, was more just something to think about

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [reply nixos/modules/services/networking/pppd.nix]: The only reason I would consider it is that adding `pppd` to `environment.systemPackages` won't add your wrapped version (which I'm guessing is needed to debug). It is probably fine not to have it

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: May be worth renaming this variable

File: nixos/modules/services/networking/pppd.nix
Comment (alexarice) [inline nixos/modules/services/networking/pppd.nix]: I think I would prefer `wantedBy = mkIf peerCfg.autostart [ "multi-user.target" ]` but your call

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: Good point!

IMO, it's still easy to debug with `systemctl restart pppd-name`, so having `pppd` or the wrapper in `systemPackages` is just more noise for nothing. And it's easier to add it later if we're wrong, than to remove it and potentially break someone who depends on it. So I propose we start without adding to `systemPackages`, and see how that goes.

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: I'll think about it some more tomorrow. In the meantime, I think it's easy to switch from a text config to structured later and deprecate this field.

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: I'll look around tomorrow for a way to build the options search engine locally and check for this.

**Review by alexarice — CHANGES_REQUESTED** [review]
Sorry I didn't notice this earlier. You haven't added you're module to `module-list.nix` so it won't be available.

**Comment by danderson** [comment]
Doh! I've been testing on my router with an import directly in configuration.nix, that's why. Added to module-list.nix. Thanks for catching that!

**Review by alexarice — APPROVED** [review]

**Comment by flokli** [comment]
@danderson can you also add a small nixos test with 2 vms ping each other through a ppp(d) connection?

**Comment by danderson** [comment]
Sure! I've never set up a pppoe+pppd server, so it might take me a bit, but I'm working on it.

**Comment by danderson** [comment]
I've added a test as requested. One VM runs a PPPoE server, and the other uses this pppd module to connect as a client. Test verifies bidirectional ping reachability through the PPP link. PTAL?

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: After reviewing the RFC, I believe there is no obvious mapping between pppd configs and nix. The format is just too weird and irregular.

However, per the RFC's recommendations for "unsuitable formats", I changed the type of `config` to `types.lines`, so at least specifying additional configuration works reasonably.

File: nixos/tests/pppd.nix
Comment (flokli) [inline nixos/tests/pppd.nix]: I assume this is not a real password :-)

**Comment by flokli** [comment]
@GrahamcOfBorg build nixosTests.pppd

**Review by flokli — APPROVED** [review]

File: nixos/modules/services/networking/pppd.nix
Comment (flokli) [inline nixos/modules/services/networking/pppd.nix]: `pppd-wrapped` can be removed, by setting `"${peerCfg.name}".environment` - we don't use pppd-wrapped outside the systemd unit anyway.

File: nixos/tests/pppd.nix
Comment (danderson) [reply nixos/tests/pppd.nix]: You assume correctly :) It's Flynn's password from [Tron](https://en.wikipedia.org/wiki/Tron). It's a classic "I need a quick password for testing" :P

File: nixos/modules/services/networking/pppd.nix
Comment (danderson) [reply nixos/modules/services/networking/pppd.nix]: Good call! Initially I wasn't sure if I wanted pppd in systemPackages, but we decided we only care about systemd units, so this is much cleaner.

Done. Nixos test still passes, and my home internet still works.

**Comment by flokli** [comment]
@GrahamcOfBorg test pppd

**Comment by danderson** [comment]
Anything else we need before merging this? The PR history says I requested a review from @Infinisil . I don't _think_ I did? But I can't edit the reviewer list, and I welcome more eyes on this if that's the next step :)

**Comment by flokli** [comment]
@danderson Infinisil is marked as a codeowner for the module system.

I wanted to await the nixos test results from CI, and forgot about it.

Thanks for the ping!

**Pull request merged**
